[Xcode] Use generate-platform-args in DerivedSources Makefiles

emw-apple opened the pull request:
#### f0528cd0af232932e7e4307e67e50e7ce2ec8c30
<pre>
[Xcode] Use generate-platform-args in DerivedSources Makefiles
<a href="https://bugs.webkit.org/show_bug.cgi?id=302332">https://bugs.webkit.org/show_bug.cgi?id=302332</a>
<a href="https://rdar.apple.com/164489420">rdar://164489420</a>

Reviewed by Alexey Proskuryakov.

The .resp files representing active wtf/Platform.h features are more
accurate because they support indirection. Replace the older Makefile
pattern of calling clang -dM directly and parsing its output.

For projects which do not yet run generate-platform-args during their
&quot;Derived Sources&quot; action, add it as a script phase. For projects which
do, reorder so that it runs before the Makefile.

JavaScriptCore and WebCore run source generation as part of their
installhdrs actions, because they install SPI headers which are
generated. For these targets, invoke generate-platform-args
unconditionally. For others, skip it in installhdrs.

As a drive-by cleanup, remove &quot;*.i386.resp&quot; files that in dependency
lists. watchOS Simulator was the last platform to drop support for
deploying to i386 targets.

* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/generate-platform-args:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj: Move
  generate-platform-args invocation to WebCore.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: In addition to the
  above, reorder when PAL runs in single-project builds relative to
  source generation.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make: In addition to the refactoring
  changes, remove FEATURE_AND_PLATFORM_UNDEFINES, which was not being
  used.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Tools/DumpRenderTree/DerivedSources-input.xcfilelist:
* Tools/DumpRenderTree/DerivedSources.make:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/project.pbxproj:

* Tools/DumpRenderTree/Scripts/generate-derived-sources.sh:
* Tools/Scripts/webkitpy/generate_xcfilelists_lib/generators.py:
(BaseGenerator._generate_derived): Makefiles now depend on a file in the
DerivedSources directory which is *not* built by Make. Add a clause to
the xcfilelist processor to handle input dependencies on DerivedSources
paths. This is a bit weird, since the original design was for
DerivedSources.make to be solely responsible for generating *all* build
inputs. However, we&apos;ve intended to replace Makefiles with Xcode-native
build phases over time, and this change furthers that effort.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/DerivedSources-input.xcfilelist:
* Tools/WebKitTestRunner/DerivedSources.make:
* Tools/WebKitTestRunner/Scripts/generate-derived-sources.sh:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/303053@main">https://commits.webkit.org/303053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6e896d27894c8160191d5b8fdb5e1b83e0a993e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41875 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138358 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/82580 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7a0e058b-b7dc-4e66-b5fa-77594d52ec87) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132787 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3222 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3082 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99747 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/67550 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1ce52cad-1f50-4888-8b3a-9ba5943ce2fe) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133862 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/2322 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117220 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/80450 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6699e9af-a677-4f0d-8645-9bd0abde2faf) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/130265 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2239 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35352 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/81603 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/122933 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110829 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140848 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129369 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2983 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2688 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/108261 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3029 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/113551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108215 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27550 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2276 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31977 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56020 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3051 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/66446 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162386 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2872 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/40570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3072 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2981 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
I let a recent version of Xcode make these changes, and given that it's restoring an older version of the format, it seems like it should be fine to commit.

aproskuryakov commented on Source/JavaScriptCore/DerivedSources-input.xcfilelist:
```diff
@@ -1,4 +1,5 @@
 # This file is generated by the generate-xcfilelists script.
+$(BUILT_PRODUCTS_DIR)/DerivedSources/JavaScriptCore/platform-enabled-swift-args.arm64e.resp
```
Only arm64e?

bls1999 replied:
Agreed -- @emw-apple, should this (and other instances of this same comment) be `$(WK_CURRENT_ARCH)` or `arm64` instead?

emw-apple replied:
Hm, I don't think $(WK_CURRENT_ARCH) is defined here (it's not a real build setting, just a trick that DerivedSources.make uses to pick an arbitrary architecture to use for generating sources against).

emw-apple replied:
The best I could come up with is to list all the known permutations, unfortunately. Xcode is fine with this—even though not all the paths will actually exist, it can track the dependency relationship between the `make` and `generate-platform-args` invocation.

aproskuryakov commented on Source/WebCore/DerivedSources-input.xcfilelist:
```diff
@@ -1,4 +1,5 @@
 # This file is generated by the generate-xcfilelists script.
+$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/platform-enabled-swift-args.arm64e.resp
```
Ditto.

aproskuryakov commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -48328,6 +48329,33 @@
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/platform-enabled-swift-args.arm64_32.resp",
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/platform-enabled-swift-args.arm64e.resp",
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/platform-enabled-swift-args.armv7k.resp",
+				"$(BUILT_PRODUCTS_DIR)/DerivedSources/WebCore/platform-enabled-swift-args.i386.resp",
```
Is i386 still a thing? I wish we had a way to cover all of these automagically.

emw-apple replied:
It looks like we can indeed remove i386, since watchOS simulator no longer supports deploying to 32-bit platforms.

aproskuryakov commented on Source/WebKit/DerivedSources-input.xcfilelist:
```diff
@@ -1,4 +1,5 @@
 # This file is generated by the generate-xcfilelists script.
+$(BUILT_PRODUCTS_DIR)/DerivedSources/WebKit/platform-enabled-swift-args.arm64e.resp
```
Only arm64e?

aproskuryakov commented on Source/WebKit/WebKit.xcodeproj/project.pbxproj:
```diff
@@ -21862,9 +21862,7 @@
-			platformFilters = (
-				ios,
-			);
+			platformFilter = ios;
```
I know that this is not being changed here, yet it's surprising that we have anything iOS-only.

emw-apple replied:
Me too. It turns out that these are the BEK extensions, but they are marked as iOS and visionOS only in the UI.

What's fascinating is that if I change one to being _only_ iOS, then the diff for all the _other_ targets changes to:
```diff
 /* Begin PBXAggregateTarget section */
@@ -21893,7 +21893,10 @@
                };
                5C400E6A29DB8AB500446F6F /* PBXTargetDependency */ = {
                        isa = PBXTargetDependency;
-                       platformFilter = ios;
+                       platformFilters = (
+                               ios,
+                               xros,
+                       );
                        target = 5C139DA129DB82E500D5117B /* WebContentCaptivePortalExtension */;
                        targetProxy = 5C400E6929DB8AB500446F6F /* PBXContainerItemProxy */;
                };
@@ -21909,19 +21912,28 @@
                };
                5CE4B60729CEBB760038F565 /* PBXTargetDependency */ = {
                        isa = PBXTargetDependency;
-                       platformFilter = ios;
+                       platformFilters = (
+                               ios,
+                               xros,
+                       );
                        target = 5CA5A2C929CBBA1A000F1046 /* WebContentExtension */;
                        targetProxy = 5CE4B60629CEBB760038F565 /* PBXContainerItemProxy */;
                };
                5CE4B62329CF880B0038F565 /* PBXTargetDependency */ = {
                        isa = PBXTargetDependency;
-                       platformFilter = ios;
+                       platformFilters = (
+                               ios,
+                               xros,
+                       );
                        target = 5CE4B61529CF877F0038F565 /* GPUExtension */;
                        targetProxy = 5CE4B62229CF880B0038F565 /* PBXContainerItemProxy */;
                };
                5CE4B62529CF880B0038F565 /* PBXTargetDependency */ = {
                        isa = PBXTargetDependency;
-                       platformFilter = ios;
+                       platformFilters = (
+                               ios,
+                               xros,
+                       );
                        target = 5CE4B60829CF87680038F565 /* NetworkingExtension */;
                        targetProxy = 5CE4B62429CF880B0038F565 /* PBXContainerItemProxy */;
                };
```

So I think we are seeing a special case where the visionOS setting inherits from iOS, and—for compatibility with older releases—Xcode omits visionOS when it would be superfluous.

aproskuryakov commented on Tools/DumpRenderTree/DerivedSources-input.xcfilelist:
```diff
@@ -1,4 +1,5 @@
 # This file is generated by the generate-xcfilelists script.
+$(BUILT_PRODUCTS_DIR)/DerivedSources/DumpRenderTree/platform-enabled-swift-args.arm64e.resp
```
Only arm64e?

aproskuryakov commented on Tools/WebKitTestRunner/DerivedSources-input.xcfilelist:
```diff
@@ -1,4 +1,5 @@
 # This file is generated by the generate-xcfilelists script.
+$(BUILT_PRODUCTS_DIR)/DerivedSources/WebKitTestRunner/platform-enabled-swift-args.arm64e.resp
```
Only arm64e?

emw-apple committed:
[Xcode] Use generate-platform-args in DerivedSources Makefiles
https://bugs.webkit.org/show_bug.cgi?id=302332
rdar://164489420

Reviewed by Alexey Proskuryakov.

The .resp files representing active wtf/Platform.h features are more
accurate because they support indirection. Replace the older Makefile
pattern of calling clang -dM directly and parsing its output.

For projects which do not yet run generate-platform-args during their
"Derived Sources" action, add it as a script phase. For projects which
do, reorder so that it runs before the Makefile.

JavaScriptCore and WebCore run source generation as part of their
installhdrs actions, because they install SPI headers which are
generated. For these targets, invoke generate-platform-args
unconditionally. For others, skip it in installhdrs.

As a drive-by cleanup, remove "*.i386.resp" files that in dependency
lists. watchOS Simulator was the last platform to drop support for
deploying to i386 targets.

* Source/JavaScriptCore/DerivedSources-input.xcfilelist:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/generate-platform-args:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj: Move
  generate-platform-args invocation to WebCore.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj: In addition to the
  above, reorder when PAL runs in single-project builds relative to
  source generation.
* Source/WebGPU/WebGPU.xcodeproj/project.pbxproj:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make: In addition to the refactoring
  changes, remove FEATURE_AND_PLATFORM_UNDEFINES, which was not being
  used.
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj:
* Tools/DumpRenderTree/DerivedSources-input.xcfilelist:
* Tools/DumpRenderTree/DerivedSources.make:
* Tools/DumpRenderTree/DumpRenderTree.xcodeproj/project.pbxproj:

* Tools/DumpRenderTree/Scripts/generate-derived-sources.sh:
* Tools/Scripts/webkitpy/generate_xcfilelists_lib/generators.py:
(BaseGenerator._generate_derived): Makefiles now depend on a file in the
DerivedSources directory which is *not* built by Make. Add a clause to
the xcfilelist processor to handle input dependencies on DerivedSources
paths. This is a bit weird, since the original design was for
DerivedSources.make to be solely responsible for generating *all* build
inputs. However, we've intended to replace Makefiles with Xcode-native
build phases over time, and this change furthers that effort.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/DerivedSources-input.xcfilelist:
* Tools/WebKitTestRunner/DerivedSources.make:
* Tools/WebKitTestRunner/Scripts/generate-derived-sources.sh:
* Tools/WebKitTestRunner/WebKitTestRunner.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/303053@main
